### ISSUE TITLE
MAG-1 — Magnetic MACE on the new architecture

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1584]
**Depends on:** ARCH-3 (#1562), ARCH-5 (#1633), CORE-1 (#1555), TRN-2 (#1576), FT-3 (#1582), DEP-1a (#1634) · **Blocks:** —

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** The `MagneticMACE` family (PR #1244, commit `7e0af5e`) is a **KEEP**: models with per-atom magnetic moments as input, magnetic-moment prediction, and **magnetic forces `dE/dm`** (the energy gradient w.r.t. the input magmoms). In legacy it is a parallel class family bolted on across the whole stack: `MagneticMACE` (`mace/modules/extensions.py:1434`), `MagneticScaleShiftMACE` (`:1712`), `MagneticSCFMACE` (`:1976`, a self-consistent-field wrapper that iterates the magmoms to a fixed point), spin-orbit-coupled-density interaction blocks (`mace/modules/blocks.py:1602,1632,1785`), a separate `MagneticMACECalculator` (`mace/calculators/mace.py:1006-1551`, 583 lines, result keys `MACE_magmoms` + magnetic forces), SO(3) magmom data augmentation implemented as a loader wrapper (`create_random_rotation_loader`, `mace/data/augmentation.py:108`, wired ad hoc in `mace/tools/train.py:220-225` from `run_train.py:1053`), **13 train flags** — the seven in the "magnetic mace specific arguments" section (`mace/tools/arg_parser.py:1173-1221`: `--num_mag_radial_basis_one_body`, `--m_max`, `--max_m_ell`, `--num_mag_radial_basis`, `--use_magmom_one_body`, `--train_one_body_contribution`, `--data_aug_magmom`) plus `--compute_magforces` (`:444`), `--magmom_key` (`:701`) / `--magforces_key` (`:707`), `--magforces_weight` (`:803`) and the `--swa_magforces_weight` / `--stage_two_magforces_weight` alias pair (`:809-815`, one `dest`); **2 eval-CLI flags** `--magmom_key` / `--return_magforces` writing `dE/dm` to `atoms.arrays` (`mace/cli/eval_configs.py:116,:123,:427`); **2 default property keys** `REF_magmom` / `REF_magforces` (`mace/tools/default_keys.py:17-18`) — these extend the *data contract* every labelled XYZ depends on; **magforces metrics in `MACELoss`** (`mae_magf`, `rel_mae_magf`, `rmse_magf`, `rel_rmse_magf`, `q95_magf`, `mace/tools/train.py:646-650,:771-779`) and in the stress log line (`mace/tools/train.py:78-84`); a `[magnetic]` extra pinning `sphericart-torch==1.0.9` **and declaring external `torch-geometric`** (`setup.cfg:80-82`); a `magnetic` capability marker (`pyproject.toml:29`, probe `sphericart` in `tests/conftest.py:66`) with its own `ci-extensions.yaml` job; and foundation-hyperparam inheritance (`inherit_magnetic_hyperparameters_from_foundation`, `mace/tools/multihead_tools.py:514`, called at `run_train.py:213-219`, with the weight-transfer side in `load_foundations_elements_magnetic`, `mace/tools/finetuning_utils.py:470`).

Two facts worth knowing before picking this up. **`MagneticSCFMACE` is not CLI-reachable** — `--model` offers ten names and only `MagneticScaleShiftMACE` among them (`arg_parser.py:137-148`); the SCF variant is constructed programmatically (`MagneticSCFMACE(model=…, n_scf_step=2)`), which is precisely why it maps onto a model-transform rather than a model class. And **the existing coverage is now substantial**: `tests/extensions/magnetic/test_magmace.py` holds **31 test functions** covering e2e train, eval, the SCF run, config round-trip, foundation-hyperparam inheritance, **8** `resolve_m_max` cases, rotation equivariance and inversion parity, optimizer-parameter registration, the magnetic calculator's backend/dtype/committee behaviour and `check_state` magmom tracking, the SCF wrapper's attribute delegation (`:849`) and its refusals (`--return_magforces` at `:741`, Hessian at `:922`), plus `tests/unit/test_augmentation.py` (rotation consistency of magmom/magforces, uniformity of the sampled orientations). What is genuinely missing is only the *numerical* golden (P0-3b), so this ticket's parity work has a solid behavioural floor under it.

This ticket re-lands the feature on the v1 architecture **declaratively** — as data + observables + transforms, not a forked class family. It is also the designated real-world exercise of the "arbitrary spherical-tensor observable declared purely in config" must-have that GATE-3 (#1585) checks.

**Interface & constraints:**

1. **Magmom is a per-atom input feature**, entering through the same input-feature mechanism ARCH-3 (#1562) builds for graph-level total charge / spin / electronic temperature (`--embedding_specs`, `arg_parser.py:743`) — the per-atom analogue, carried in the batch schema, not a new model-class fork.
2. **Magnetic moment is an observable**: a per-atom vector (`1o` irrep) declared via CORE-1 (#1555)'s `ObservableSpec` (`name`, `irreps`, `per_atom`, `units`, `normalization`, `default_loss_weight`), with ARCH-3 (#1562)'s observables→automatic-head-creation producing the head — zero edits to core model files.
3. **`dE/dm` is a derivative of the energy w.r.t. the magmom input**, generated and auto-named by CORE-1 (#1555)'s derivative machinery (the same grammar that yields `d_<q>_d_pos`/`d_<q>_d_cell`, extended over the magmom input); its loss term comes from TRN-2 (#1576)'s observable-spec-generated losses, and its metrics are the five `*_magf` entries TRN-3 (#1577) ports. Sign and units must match the legacy magnetic forces.
4. **SO(3) magmom augmentation is a registered TRN-2 (#1576) data transform** (rotate positions/cell and magmoms consistently), replacing the legacy loader-wrapping hack — selectable from config like any other transform. Develop's `Random3DRotation` (`mace/data/augmentation.py:24`) already documents the copy semantics it relies on (`BaseTransform.__call__` shallow-copies before `forward`); the v1 transform must be explicit about that rather than inheriting it from a third-party base class.
5. **`MagneticSCFMACE` becomes a v1 model + the TRN-2 (#1576) model-transform hook** (SCF fixed-point solve with implicit differentiation) — the same hook contract ELEC-3 (#1593)'s FixedPointSCF family will reuse; MAG-1 (#1584) consumes the hook, TRN-2 (#1576) builds it. Develop's two refusals are the floor, and in v1 they must be structural rather than special-cased: `mace/cli/eval_configs.py:197-202` refuses `--return_magforces` for an SCF-wrapped model because the wrapper's `forward` takes no `compute_magforces`, and the Hessian path refuses it likewise (pinned at `test_magmace.py:922`). A hook that cannot differentiate through its fixed point must be rejected at build time for force training, not at call time.
6. **`[magnetic]` extra**: keep it, pinning **sphericart-torch** — and evaluate sphericart as an *accelerated* spherical-harmonics override for the kernel-backend layer (it is fast, CPU+GPU, differentiable to second order). The native plain-torch spherical harmonics remain the reference; sphericart is optional, never required. Legacy already has a working precedent to read: `SHModule` (`mace/modules/extensions.py:1357`) wraps `sphericart.torch.SolidHarmonics` and is a live non-e3nn spherical-harmonics path in-tree. `ChebyshevBasisGeneral` (`:1380`) sits next to it and is an ordinary radial basis reachable from `--radial_type chebyshev` (`arg_parser.py:153-158`) — it belongs with the ARCH-1 (#1559) bases, not with the magnetic models.
7. **The two magnetic interaction blocks are `Density` variants** (`MagneticRealAgnosticSpinOrbitCoupledDensityInteractionBlock`, `mace/modules/blocks.py:1632`; the `Residue` variant at `:1785`, both over `MagneticInteractionBlock` at `:1602`). They therefore inherit whatever ARCH-2 (#1561) does for the Density interaction family — which is **not** optional there, since MACE-MP-0b2 (S/M/L) and MACE-mh-0 are built on the plain Density blocks. Port the magnetic pair as a spin-orbit-coupled specialization of that same backbone, not as an independent block family.
8. **Dependency collision to resolve, not inherit**: the `[magnetic]` extra declares **external `torch-geometric`** (`setup.cfg:82`), and `mace/data/augmentation.py` imports `torch_geometric.transforms.BaseTransform` behind a `find_spec` guard with a stub fallback (`:8-21`) *while also* importing the vendored `mace.tools.torch_geometric.dataloader` (`:6`). v1 deletes the vendored copy and builds collation without torch_geometric at all, so the ported augmentation transform must depend on **neither**: implement it against the v1 transform interface, and drop `torch-geometric` from the extra. Flag this in the PR so the retirement track sees the extra is clean.
9. **Calculator + CLI**: the magnetic path folds into the **one** DEP-1 (#1583)/DEP-1a (#1634) calculator (magmoms in via ASE, `MACE_magmoms` + magnetic forces out, magmom-aware `check_state` per `MagneticMACECalculator.check_state`/`_magmoms_equal`, `mace/calculators/mace.py:1267-1312`), and `mace eval` grows the magnetic-force path (magmom input key, `dE/dm` written to `atoms.arrays` at `eval_configs.py:427`) — succeeding the legacy `--magmom_key` / `--return_magforces` eval flags. Legacy ships this as a **separate `MagneticMACECalculator` class** (583 lines duplicating load, batch construction, committee and descriptor logic); in v1 a second calculator class would be the fork this ticket exists to remove. The magmom-equality half of `check_state` is real behaviour, not duplication, and must survive.
10. **Foundation-hyperparam inheritance**: a magnetic model can be built on top of a foundation model, inheriting its hyperparameters (`inherit_magnetic_hyperparameters_from_foundation`, `multihead_tools.py:457`) with weights transferred by the magnetic slicing path (`finetuning_utils.py:405`) — in v1, options on FT-3 (#1582)'s single element-transfer operation, not a third transfer function.
11. **A derivative w.r.t. an input feature is a new differentiation target.** The observable grammar covers derivatives w.r.t. positions and cell only; magnetic forces are `dE/dm`, a derivative with respect to a *declared per-atom input feature*. This ticket owns extending both the observable spec and the derivative engine with that third target — and that extension is the one permitted core-file change here, since the zero-core-edits property applies to the magnetic-moment head and its generated loss term, not to the grammar. The canonical observable name is **`magforces`**, matching the `REF_magforces` data key, the `--magforces_weight` observable weight, the `output_args["magforces"]` flag legacy threads into every step function (`mace/tools/train.py:443-444,:516-517`) and the legacy eval output; record its sign convention and units against the legacy magnetic forces.
12. **`m_max` is a typed field with two legacy input forms.** `--m_max` accepts either a dict literal mapping atomic number → saturation (unlisted elements default to 1.0) or a space-separated float list ordered by `z_table.zs` (`arg_parser.py:1180-1190`, resolved by `resolve_m_max` in `mace/tools/scripts_utils.py`, 8 pinned cases in `test_magmace.py:337-394`). v1 keeps the dict form as the config shape and the ordered-list form as an accepted legacy input in `from_namespace`, with the same defaulting, the same wrong-length error, and the same "extra dict keys are ignored" behaviour.
13. **Property keys and weights**: `REF_magmom` / `REF_magforces` join the CORE-3 (#1557) default-key set (a data contract — existing labelled datasets use these names), and `--magforces_weight` / `--stage_two_magforces_weight` become an observable weight and a per-stage schedule entry respectively, like every other observable in TRN-2 (#1576). The `--swa_magforces_weight` option-string spelling is **not** carried over (it shares the `dest`, so nothing is lost).
14. **Golden**: P0-3b trains and commits a tiny `MagneticScaleShiftMACE` (fixed seed, fp64) and snapshots energy + magnetic moments + `dE/dm` on non-periodic magmom fixtures → `references/tiny_magnetic_e3nn_cpu_fp64.json`, `magnetic`-marked. **MAG-1 (#1584) reproduces this golden at the fp64 tolerance row**; the golden is never regenerated here.

**Task:**
1. Declare the magmom per-atom input feature in the batch schema and the magnetic-moment observable (`1o`, per-atom) + its `dE/dm` derivative in observable config; verify ARCH-3 (#1562) auto-creates the head with no core-file edits beyond the derivative-target extension (constraint 11).
2. Port the Magnetic model family (spin-orbit-coupled-density interaction pattern from `mace/modules/blocks.py:1495+`, models from `mace/modules/extensions.py:1428+`) as v1 model-registry entries; `MagneticSCFMACE` via the TRN-2 (#1576) SCF/implicit-differentiation hook, with the two refusals expressed structurally.
3. Register the SO(3) magmom augmentation as a TRN-2 (#1576) transform (config-selectable), with an equivariance test: rotating inputs (positions + magmoms) rotates the predicted moments and `dE/dm` accordingly; port `tests/unit/test_augmentation.py`.
4. Keep the `[magnetic]` extra (sphericart-torch, drop torch-geometric); evaluate sphericart as an optional accelerated spherical-harmonics override and record the outcome in the PR (reference implementation unchanged either way).
5. Add the magnetic path to the DEP-1 (#1583)/DEP-1a (#1634) calculator (not a second calculator class), preserving magmom-aware `check_state`, and the magnetic-force path in `mace eval`.
6. Implement foundation-hyperparam inheritance and the magnetic weight transfer as options on FT-3 (#1582)'s element-transfer operation; port the `test_magmace.py` inheritance case.
7. Land `REF_magmom`/`REF_magforces` in the CORE-3 (#1557) default-key set, `resolve_m_max`'s two input forms in `from_namespace`, and the magforces weight/per-stage schedule in the TRN-2 (#1576) loss config.
8. Parity: reproduce `references/tiny_magnetic_e3nn_cpu_fp64.json` (energy, magnetic moments, `dE/dm`) through the v1 stack.

**Out of scope:** building the TRN-2 (#1576) model-transform hook itself (TRN-2 (#1576)); training/committing the golden (P0-3b); deleting the legacy Magnetic classes (retirement track); the charge-based SCF families (ELEC-3 (#1593)); the non-magnetic model families on the calculator (DEP-1a (#1634)).

**Acceptance criteria:**
- [ ] v1 reproduces `references/tiny_magnetic_e3nn_cpu_fp64.json` — energy, per-atom magnetic moments, and `dE/dm` — at the fp64 tolerance row.
- [ ] The magnetic-moment observable and its `dE/dm` derivative are declared purely in config; head and loss are generated with zero core-file changes beyond the derivative-target extension (the GATE-3 (#1585) must-have, demonstrated on this feature).
- [ ] SO(3) augmentation transform is registered and config-selectable; the equivariance test (rotated magmom inputs → rotated moment/`dE/dm` outputs) and the orientation-uniformity test pass.
- [ ] `MagneticSCFMACE` converges its fixed point and gradients flow through the implicit-differentiation hook (gradcheck on a tiny system); an SCF model that cannot differentiate its fixed point is refused at build time for force training, and `mace eval`'s magforce path refuses it with the develop message rather than a `TypeError`.
- [ ] The one calculator returns magmoms and magnetic forces through ASE, tracks magmom changes in `check_state`, and there is no second calculator class; `mace eval` writes `dE/dm` to `atoms.arrays`.
- [ ] `resolve_m_max`'s eight pinned cases pass through the v1 config path (dict form, ordered-list form, single-float broadcast, `None`, wrong length, extra keys, numpy atomic numbers).
- [ ] All magnetic tests carry the `magnetic` capability marker and skip cleanly when the `[magnetic]` extra is absent; the `ci-extensions.yaml` `magnetic` job stays green with `MACE_REQUIRE_CAPS=magnetic`.
- [ ] The `[magnetic]` extra no longer declares `torch-geometric`, and nothing on the magnetic path imports it (real or vendored) — checked by the import-direction guard.
- [ ] `REF_magmom` and `REF_magforces` parse from a labelled XYZ through the v1 key convention with the same default names as legacy.

**Inventory gaps assigned here:**
- [ ] The numeric golden is the only real gap (P0-3b produces it, MAG-1 (#1584) consumes it); the behavioural floor is `tests/extensions/magnetic/test_magmace.py` (31 funcs) plus `tests/unit/test_augmentation.py` (2 funcs).
- [ ] `--use_magmom_one_body` / `--train_one_body_contribution` / `--num_mag_radial_basis_one_body` — the one-body magmom contribution and whether its coefficients are optimized; no test asserts the non-default combination.

**Verify:**
```bash
pip install -e "packages/mace-torch[magnetic]"
python -m pytest packages/mace-torch/tests/models/test_magnetic.py -v      # equivariance, SCF gradcheck, calculator
python -m pytest tests/parity -m "not slow" -k magnetic                    # vs references/tiny_magnetic_e3nn_cpu_fp64.json
MACE_REQUIRE_CAPS=magnetic python -m pytest tests/extensions/magnetic tests/unit/test_augmentation.py -v
python -m pytest packages/mace-torch/tests -m magnetic                     # marker + skip-without-extra discipline
```
**Review focus:** sign/units of `dE/dm` against the legacy magnetic forces; equivariance of the `1o` observable under the augmentation transform; the implicit-differentiation correctness of the SCF hook usage and that its two refusals are structural; and that magmom enters as data + config (no parallel model-class family and no second calculator class recreated).

---
